### DISCUSSION TITLE
Myelin separation

### DISCUSSION THREAD

**Discussion by mcib3d** [neuropoly/axondeepseg#661] [Show and tell]
Hi and thanks for this wonderful tool,

I would eventually perform more analysis on the detected objects axons and myelins. I can  easily get a labelled image with the axons since they are separated. Sometimes two myelins touch, so my questions are : 

- how are they separated in the analysis ? It is a simple watershed ?
- can we get access to the _hidden_ separated myelins ?

Many thanks.

Thomas

**Comment by mathieuboudreau** [comment]
Hi Thomas!

>can we get access to the hidden separated myelins ?

By this, do you mean the myelin that's "removed" to draw the line between the touching myelin?

**Reply by mcib3d** [reply]
Hi Mathieu,

Thanks for your reply, yes anything that will allow label individuals myelins, so I guess the pixels _removed_ to separate the myelins should work.

**Reply by mathieuboudreau** [reply]
>yes anything that will allow label individuals myelins

Oh, for this purpose we generate IDs for each axon-myelin objects when the morphometrics file is generated (see doc: https://axondeepseg.readthedocs.io/en/latest/documentation.html#morphometrics), the (x0, y0) values in the file are the locations of the "center" of the axon-myelin objects, and other image files are generated with an overlay of the ID number of the axons to help you identify these visually:

![image_axonmyelin_index](https://user-images.githubusercontent.com/1421029/177558263-b33282b5-6f38-416f-8b5e-720f39884428.png)

**Reply by mcib3d** [reply]
Hi Mathieu,
I would like to create a _classical_ label image where each object (here myelin for instance) has its own gray value. In this image, myelin 91 and 81 are touching, I would like to  get access to all the pixels of the different objects, like in this example : [Segmentation](https://mcib3d.frama.io/3d-suite-imagej/plugins/Segmentation/3D-Iterative-Segmentation/)
You do not need to generate this kind of image to compute morphometrics ?

**Comment by hermancollin** [comment]
Hm @mcib3d I see what you mean. You would like to have access to the instance segmentation that we obtain during the postprocessing step (morphometrics) instead of the semantic segmentation.

I briefly worked on a tool to output the instance segmentation (with different colors so you could see the delineation well). I obtained something like this:
![sub-nyuMouse36_sample-0001_TEM_seg-instance](https://user-images.githubusercontent.com/83031821/177560899-0efd7138-40db-473d-ba16-402a6a13239c.png)

If you are interested in this feature, I could implement it so that you can save the instance segmentation as well.

**Reply by hermancollin** [reply]
Also I guess I would need to work on this a little more so that it uses only 4 colors (mathematical minimal value for 2D maps).

**Reply by mcib3d** [reply]
Yes exactly,  instance segmentation  :-)  . If this could be available (as  an option I guess), this would be great, so you can then do your own analysis on the objects.

**Reply by mcib3d** [reply]
Ideally an instance segmentation should be an 8-bit (or better 16-bit) grey-level image, with each object having its own grey value, you can add colours later on with some LUT.

**Reply by Stoyan-I-A** [reply]
@hermancollin You could use the `get_watershed_segmentation` in PR #651 for this

**Reply by hermancollin** [reply]
I agree that this is a very important feature for the user to review his results. I will add an option to save the instance seg ASAP and get back to you.

**Reply by hermancollin** [reply]
> @hermancollin You could use the `get_watershed_segmentation` in PR #651 for this

thats what I did 😏

**Reply by mcib3d** [reply]
Sounds perfect, merci !

**Reply by hermancollin** [reply]
> Ideally an instance segmentation should be an 8-bit (or better 16-bit) grey-level image, with each object having its own grey value, you can add colours later on with some LUT.

@mcib3d Unfortunately, I get your point but I've often seen images where the compression messes up the exact grayscale values. Additionnally, we sometimes work with images with 256+ axons so 8 bit wouldn't be enough (not even sure about 512).

We will discuss this internally before proceeding. If you have any more input, feel free to let us know.

**Comment by hermancollin** [comment]
Hi @mcib3d,
In #667, I'm implementing the feature we discussed here. One issue I am encountering is that the colorization computation is very slow for large images. I'm trying to solve this issue but could you give me an estimate of how many axons you will have per image? A few dozens? Hundreds? Thousands?

**Comment by mcib3d** [comment]
Hi @hermancollin,

Thanks for your implementation, I do not know how you store your myelin objects, but it seems to me that colouring is more complex than just labelling, with a gray image, label myelin 1 with value 1, ...
To answer your question, I think, including small detections, we may have up to 100 objects.
Best,
Thomas

**Reply by hermancollin** [reply]
If you have 100-200 fibers, the current implementation should work just fine. It's for larger images that it becomes a problem.

**Comment by mcib3d** [comment]
At the end I will create a labelled gray image, it can be done starting from a coloured image, as illustrated here : 


![segAxon](https://user-images.githubusercontent.com/8034509/180380340-c175bb96-dc4f-4a7d-b8cc-c566e6e12ab4.png)

**Reply by hermancollin** [reply]
Yes, I understand you want to label every instance with a different value, but from our perspective, we often deal with images with thousands of axons. In that case, you would need 16-bit images to encode all instance IDs, which takes way more space than a colored image. However, I understand how the current instance map output is not really convenient if you want to extract information from it. Maybe the best solution for now would be to output a 16-bit image and just directly print the instance segmentation onto it, like you said:
> just labelling, with a gray image, label myelin 1 with value 1

**Comment by mcib3d** [comment]
Hi @hermancollin,

Yes, an instance labelled 16-bit is exactly what we need for further analysis. Note that I usually save labelled images as .zip to gain space. 
Thanks and happy to test it when it is released.
Best,
Thomas

**Comment by hermancollin** [comment]
@mcib3d I will take care of adding the 16-bit instance segmentation image this week. In the meantime, you can already use the colorization feature implemented in #667. You just need to `git checkout master; git pull` in the axondeepseg directory.
Using `axondeepseg_morphometrics` with the `-c` flag will now output a colorized image, as per https://axondeepseg.readthedocs.io/en/latest/documentation.html#colorization. I'll ping you when the 16-bit image feature is implemented.

**Reply by mcib3d** [reply]
Hi @hermancollin ,
Perfect, it works nice, I will try to do my own instance segmentation from this and see how it goes. Many thanks. 
Thomas
